### DISCUSSION TITLE
What is sideeffect

### DISCUSSION THREAD

**Discussion by FuXiii** [skaarj1989/FrameGraph#2] [General]
When I read `test` in `test.cpp`
```CXX
builder.setSideEffect();

FrameGraph::Builder &FrameGraph::Builder::setSideEffect() {
  m_passNode.m_hasSideEffect = true;
  return *this;
}
```
and
```CXX
FrameGraphResource FrameGraph::Builder::write(FrameGraphResource id) {
  assert(m_frameGraph.isValid(id));
  if (m_frameGraph._getResourceEntry(id).isImported()) setSideEffect();
...
}
```

What is the `m_hasSideEffect ` use for?

And the `FrameGraph-Example` just a little complex, I am just a fresh hand. Can you update an example about how to draw a box with index buffer? :D Thanks~

**Comment by skaarj1989** [comment]
`setSideEffect` ensures that a pass is not culled during the compilation phase. (writing to imported resource counts as side-effect).

Moving to FrameGraph is a big paradigm shift. It might be overwhelming at first but once you understand it you won't look back ;)

I encourage you to watch/read:
- https://www.gdcvault.com/play/1024045/FrameGraph-Extensible-Rendering-Architecture-in
- https://www.gdcvault.com/play/1024612/FrameGraph-Extensible-Rendering-Architecture-in

**Comment by FuXiii** [comment]
Oh! I have generally finished reading your project and it wrote very well and clearly, Thank you very much~ :D. 

---
When I try to read `FrameGraph` in `Filament`. `FrameGraph` looks like rebuild every frame? If I have many many `Node` and `Resource` in one rendering frame, if some `Node` not change but I must rebuild it at next new frame? Is that will waste performance？ Is that really necessary to rebuild `FrameGraph` every frame?

---
If I create a `PassNode` and create an `TextureNode` for read, there have `assert` code in `FrameGraph::Builder::read(...)`:
```CXX
FrameGraphResource FrameGraph::Builder::read(FrameGraphResource id) {
  assert(m_frameGraph.isValid(id));
  return m_passNode._read(id);
                      |_ assert(!creates(id) && !writes(id));
}
```
It looks like `FrameGraph` not support `PassNode` to create `ResourceNode` for read.

---

If I want create an `TextureNode` for  `ColorDrawPass` and it will sampled in fragment shader. 
1. How I binding the shaders in `PassNode`
2. Resource are `create` and `destroy` by `ResourceEntry::Concept or Model` . How to initial the texture's color pixel? If I `import` this texture into `FrameGraph` how to do it? the `test` in `TEST_CASE("Imported resource", "[FrameGraph]")` the `id` for the import already existed
```CXX
 constexpr auto kBackbufferId = 777;
 FrameGraph fg;
 const auto backbuffer = fg.import("Backbuffer", {1280, 720}, FrameGraphTexture{kBackbufferId});
```

In really case the code maybe looks like:
```
Texture texture = ImageLoader::Load("./SomePicture.png");
FrameGraph fg;
fg.import(..., texture );
```
Is that right?

or just create an `HostWritePass` for host texture memory write? but I think this just a little odd.

![image](https://user-images.githubusercontent.com/40132375/171971182-da67ab79-f1cc-41e3-be6f-96dd5d31b818.png)

**Comment by skaarj1989** [comment]
> Is that really necessary to rebuild FrameGraph every frame?

Yes, you need to rebuild the FrameGraph every frame.

> (...) Is that will waste performance?

Here I have a FrameGraph with 38 passes and 51 resources:

![image](https://user-images.githubusercontent.com/34756939/171987039-6c6bf019-3006-4da4-a021-5e805c1c8d0d.png)
![image](https://user-images.githubusercontent.com/34756939/171987908-e1114c05-ab90-4605-a356-8e0ecc6f8ff6.png)

(measured with [Tracy](https://github.com/wolfpld/tracy) Profiler).
In my opinion, the cost is negligible.

> It looks like FrameGraph not support PassNode to create ResourceNode for read.

Do you want to create a resource in a pass and read it immediately in the same pass? That will give you a garbage data.
Usually, a pass creates a resource (writes to it) and another pass reads from it.

> How I binding the shaders in PassNode

It's up to you. Pass whatever you need to `ctx` in an execution callback.

```cpp
struct PassData {};
fg.addCallbackPass<PassData>("SimplePass",
  [&](FrameGraph::Builder &builder, PassData &data) {
    // ...
  },
  [=](const PassData &data, FrameGraphPassResources &resources, void *ctx) {
    auto &rc = *static_cast<RenderContext *>(ctx);
    rc.bindPipeline(/* whatever, could be a member variable */);
    // ...
  }
);
```

Here you have an example:
https://github.com/skaarj1989/FrameGraph-Example/blob/7867df8e3146acfffaf41173d5776b6ac265fd50/src/WorldRenderer.cpp#L415

---

1. Don't use a FrameGraph to create a resource that is initialized once (like mesh vertex buffer or a material texture).
2. Don't import everyting into a FrameGraph.
 
[Here](https://github.com/skaarj1989/FrameGraph-Example/blob/master/src/Passes/SSAO.cpp), the SSAO pass is the owner of a noise texture and a kernel uniform buffer. Both are initialized by the constructor and then captured by the execution callback.

---

To create a resource, the FrameGraph class uses given `allocator` (transient resource manager/system).

```cpp
void execute(void *context = nullptr, void *allocator = nullptr);
```

Take a look at [here](https://github.com/skaarj1989/FrameGraph-Example/blob/master/src/TransientResources.cpp), for an example of a transient resource system.

**Comment by FuXiii** [comment]
Oh~ Thank you very very much~ :D
I think I need to read your `FrameGraph-Example` project .♪(^∇^*)

**Comment by FuXiii** [comment]
When to judge `ID Valid`
```CXX
bool FrameGraph::isValid(FrameGraphResource id) const {
  const auto &node = _getResourceNode(id);
  auto &resource = m_resourceRegistry[node.m_resourceId];
  return node.m_version == resource.m_version;
}
```

the code compare the `version` between `ResourceNode` and `ResourceRegistry`. Why? Why to compare the `version`?

I know if different `pass` write to the same `ResourceEntry` ,the `version` of `ResourceEntry` will plus one for each `clone` and set the `new ResourceEntry version` into the `ResourceNode version`, that will let track resource very easy, but I don't understand why use the `node.m_version == resource.m_version` to judge whether it is valid？

**Comment by skaarj1989** [comment]
It enforces a specific execution order of the render passes (prevents mistakes).

Imagine that you have a pass **P0**, that writes to a resource **R**.
Followed by a pass **P1**, that writes to **R** (version is bumped)
Then you want to read **R** in yet another pass **P2**
you could by mistake give the handle of the first write operation to **P2** (you should use the most recent one)

I'm gonna convert this issue to a discussion (since it's not really an issue).

The original question is about the `sideEffect`. If you have more questions just create a new discussion.

**Reply by FuXiii** [reply]
https://themaister.net/blog/2017/08/15/render-graphs-and-vulkan-a-deep-dive/

In this blog I remember it has cover some topic about reorder the PassNode.

please check the `Traverse dependency graph` section

**Reply by skaarj1989** [reply]
I am not going to implement pass reordering. I don't need that feature.

**Reply by FuXiii** [reply]
And in `ResourceEntry::ResourceEntry(...)`
```CXX
template <typename T>
inline ResourceEntry::ResourceEntry(uint32_t id, typename T::Desc &&desc,
                                    T &&obj, uint32_t version, bool imported)
    : m_id{id}, 
     m_concept{std::make_unique<Model<T>>(
                  std::forward<typename T::Desc>(desc), std::forward<T>(obj))}
```
the 
```CXX
T &&obj
```
It looks like not had been used.
It just use `T` to define a `T resource` in `template <typename T> struct Model`

**Reply by skaarj1989** [reply]
It's part of type erasure. I don't understand your point here.

**Reply by FuXiii** [reply]
Oh, the `type erasure` looks like a feature of  `Templete Program`, I am not very familiar with `type erasure`

**Comment by FuXiii** [comment]
Hello I am coming again.
In `FrameGraphPassConcept` this struct has declare ` virtual void operator()(FrameGraphPassResources &, void *) = 0;` , and the child class `FrameGraphPass` use this to reload `operate()` member function to invoke the `execFunction` function. 
I have not find some code like:
```Cpp
//FrameGraphPass fgp = ...;
//FrameGraphPassResources fgpr = ...;
//void* custom_data = ...;
fgp(fgpr, custom_data);
```
I only find the `std::invoke(...)` call to invoke `execFunction`, such like the calling in `void FrameGraph::execute(...)`:
```Cpp
std::invoke(*pass.m_exec, resources, context);
```
Is that means the `std::invoke(*pass.m_exec, resources, context)` will invoke the `FrameGraphPass::operator()(...)`?

**Comment by skaarj1989** [comment]
Once again it's type erasure, but this time for a function.

```cpp
FrameGraph fg;

struct PassData {};
fg.addCallbackPass<PassData>("SimplePass",
  [&](auto &builder, PassData &data) {
    // setup callback, called immediatly via fg.addCallback
  },
  [=](const PassData &data, FrameGraphPassResources &resources, void *) {
    // this is an execution callback, it's execution is deferred.
    // (see the very last line of this post)
  }
);
```

Execution callback is wrapped:
https://github.com/skaarj1989/FrameGraph/blob/326e6762dcec39e6dbe529fb1be88910f7e33377/include/fg/FrameGraph.inl#L15

```cpp
fg.execute()
```

https://github.com/skaarj1989/FrameGraph/blob/326e6762dcec39e6dbe529fb1be88910f7e33377/src/FrameGraph.cpp#L99

Here, a pass `Data` is smuggled into the execution callback:
https://github.com/skaarj1989/FrameGraph/blob/326e6762dcec39e6dbe529fb1be88910f7e33377/include/fg/PassEntry.hpp#L23

**Reply by FuXiii** [reply]
Is that means, When you call:
```CXX
std::invoke(*pass.m_exec, resources, context); 
```
this `std::invoke(...)`,it will call:
```CXX
execFunction(data, resources, context); 
```
which is define of :
```CXX
void operator()(FrameGraphPassResources &resources, void *context) override 
```

**Reply by skaarj1989** [reply]
Almost.

```cpp
std::invoke(*pass.m_exec, resources, context); 
// -> FrameGraphPassConcept::operator() -> FrameGraphPass::operator() -> wrapped execution callback
```

**Reply by FuXiii** [reply]
Oh, Thank you. will your hint, I found this in cppreference:
![image](https://user-images.githubusercontent.com/40132375/176668814-444798dd-7176-454e-b58c-ecd76fb634f0.png)

Thank you very much 👍
